src/Hooks/StrictEqualityHooks.php

```diff
@@ -126,33 +139,135 @@ private static function isEqualOrdered(Atomic $first_type, Atomic $second_type):
+            if (self::isTooComplicatedType($atomic_type)) {
+                continue;
+            }
+
+            if ($atomic_type instanceof Atomic\TCallable) {
+                continue;
+            }
+
+            if ($atomic_type instanceof Atomic\TResource) {
+                continue;
+            }
+
+            if ($atomic_type instanceof Atomic\TClosedResource) {
+                continue;
+            }
+
+            return false;
         }
 
-        if ($first_type instanceof Atomic\TBool && $second_type instanceof Atomic\TBool) {
-            return true;
+        return true;
+    }
+
+    private static function isUnionStringNotEqualOrdered(array $first_types, array $second_types): bool
+    {
+        // identical at the moment
+        return self::isUnionStringEqualOrdered($first_types, $second_types);
+    }
+
+    private static function isUnionEqualOrdered(array $first_types, array $second_types): bool {
+        $top_level_class = false;
+        foreach ($first_types as $atomic_type) {
+            if ($top_level_class === false) {
+                $top_level_class = $atomic_type;
+                continue;
+            }
+
+            if ($atomic_type instanceof $top_level_class) {
+                continue;
+            }
+
+            if ($top_level_class instanceof $atomic_type) {
+                $top_level_class = $atomic_type;
+                continue;
+            }
+
+            return false;
         }
 
-        if ($first_type instanceof Atomic\TArray && $second_type instanceof Atomic\TArray) {
-            return true;
+        if (self::isTooComplicatedType($top_level_class)) {
+            return false;
         }
 
-        if ($first_type instanceof Atomic\TKeyedArray && $second_type instanceof Atomic\TKeyedArray) {
-            return true;
+        foreach ($second_types as $atomic_type) {
+            if ($atomic_type instanceof $top_level_class) {
+                continue;
+            }
+
+            if ($top_level_class instanceof $atomic_type) {
+                $top_level_class = $atomic_type;
+                continue;
+            }
+
+            return false;
         }
 
-        if ($first_type instanceof Atomic\TKeyedArray && $second_type instanceof Atomic\TArray) {
-            return true;
+        return true;
+    }
+
+    private static function isUnionNotEqualOrdered(array $first_types, array $second_types): bool
+    {
+        // identical at the moment
+        return self::isUnionEqualOrdered($first_types, $second_types);
+    }
+
+    private static function isTooComplicatedType(Atomic $type) {
+        $too_complicated_types = array(
+            Atomic\TKeyedArray,
+            Atomic\TArray,
+            Atomic\TList,
+            Atomic\TIterable,
+            Atomic\TNamedObject,
+            Atomic\TObject
+        );
+
+        foreach ($too_complicated_types as $compare) {
+            if ($type instanceof $compare || $compare instanceof $type) {
```

Comment:
Not sure why you check in both ways. The first should be enough

Reply:
I think it will crash. instanceof expects an object in its first operand, not a class-string